### ISSUE TITLE
Mapping HLM PFTs onto FATES PFTS in reduced complexity modes

### ISSUE THREAD

**Issue by rosiealice** [NGEET/fates#631]
Our current plan for the SP and fixed biogeography (FBG) modes, is to have FATES read the HLM surface datasets and pass the PFT areal fractions into FATES, whereupon FATES will use that information to restrict the distribution of PFTs geographically. However, until now we have had no need for the FATES and HLM PFT definitions to be in any way compatible, and so this new structure raise the question of how to manage that interface.  

@ckoven @rgknox @glemieux and I had a discussion about thin on 13/04/20 which was inconclusive and I am going to try and paraphrase here for wider discussion. 

There are broadly **three scenarios**, defined here with HLM:FATES as the nomenclature. 

**1:1** -  The FATES PFT are the same as the HLM PFTs, and no mapping is required.

**many:1** - Each FATES PFT corresponds to >=1 HLM PFTs. e.g. FATES Evergreen Needleaf Tree corresponds to the boreal and temperature ENT's in the HLM. 

**1:many** - The HLM defines one PFT per biome whereas FATES uses 'n' PFTs to describe that biome. (e.g. the HLM has one BET whereas FATES has five BETs defined by hydraulic strategy).

The options we discussed for managing this were 
### 1. Invent a tool to convert the surface datasets to the same PFT format as FATES . 
I had thought this might be easiest but @ckoven pointed out that once we get into managing land use datasets the overheads would be much higher, so probably not going to go with that. 

### 2. Embed mapping into FATES that converts from HLM to FATES PFT structures
The dimensionality of the mapping is perhaps the most critical decision here. It could either have:
i) The dimensions of the HLM PFT file. And implement the many:1 case by having integers in each element corresponding to the FATES PFTs e.g.  for a case where the HLM has 12 PFTs and FATES 6, that would look something like this:
[1 1 2 2 2 3 4 4 5 5 6 6]

ii)  The dimensions of the FATES PFT file x a maximum umber of HLM PFTs that could be shared by a FATES PFT. For the many:1 case that would look something like
[1,2; 3,4,5;  6;  7,8;  9,10;  11,12] (where semi-colons delineate new columns and commas rows in columns) 

iii) The dimensions of a matrix, wherein the elements could be fractions. For the many:1 case this would simply make the above information more sparse (and, I think, would basically need to be collapsed down into something like option 1 or 2 to be used in the code anyway) 
FATES PFT:  1 2 3 4 5 6
  HLM1        [ 1 0 0 0 0 0]
  HLM2        [1 0 0 0 0 0]
  HLM3        [0 1 0 0 0 0]
  HLM4        [0 1 0 0 0 0]

For the 1:many case, the integer matrix might thus look something like this (where we are using 2 HLM PFTs to define 6 FATES PFTs):
FATES PFT:  1 2 3 4 5 6
  HLM1        [ 1 1 1 1 0 0]
  HLM2        [0 0 0 0 1 1]
  HLM3        [0 0 0 0 0 0]
  HLM4        [0 0 0 0 0 0]

In this 1:many case, to do FBG- no comp or SP modes, one would further need to define the fraction of the HLM areal fraction that was associated with each FATES PFT.   @ckoven suggested that we could have fractions populating this matrix to divide up the land area. e.g.
FATES PFT:  1       2       3       4 .    5 .   6
  HLM1        [ 0.25 0.25 0.25 0.25 0     0]
  HLM2        [0       0      0       0      0.5  0.5]
  HLM3        [0       0      0       0      0     0]
  HLM4        [0       0      0       0      0     0]

As @rgknox pointed out on the call, this would fix the areal fractions in space, but also, we probably won't, in most cases, have the capacity to actually know the areal fractions of lots of the PFT dimensions we want to represent, so this could still present a useful functionality. 

do we have any addition potential models for this (dictionaries?).  Going through this again, the matrix idea is perhaps the most flexible for all possible use cases. What do you all think?

**Comment by huitang-earth** [comment]
@rosiealice Look-up matrix is a good idea. My question is that: in SP modes, to prescribe LAI, SAI etc, is it reasonable to use the same look-up matrix to derive an area-weighted LAI for each FATES pft?

**Comment by rosiealice** [comment]
Hi @huitang-earth,

On the former, point (look up tables) I'm not sure I understand your question well enough. 

I think in FATES SP mode we will initially just use one cohort to define the PFT LAI, area, SAI, etc. -(except of course in the case that you have with potentiallly known LAIs for different cohorts). Where one PFT is made up from several HLM PFTs (ie. one evergreen needleleaf tree in FATES rather than 2-3 ENTs in the HLM) I think we would probably not represent that in FATES because FATES would have no concept of what that meant. I guess if we wanted to disaggregate the ENTs in FATES then we could add in more PFTs, unless there's a case I'm missing in which having >1 cohorts of different ENT identities would be useful??

**Comment by rosiealice** [comment]
While I am thinking about this, the 'matrix' for converting from HLM to FATES pfts should, I think, be a FATES level parameter, (and thus, @rgknox , should be on the list of 'things to be added to the interface' once this mode is finalized). 

However, putting it in the FATES file requires it to have one dimension that corresponds to the number of PFTs on the surface dataset coming from the HLM, which is, in principle, a HLM level parameter and could in principle differ between HLMs once they start to diverge in their PFT resolutions. Should we also pass a variable from the HLM to FATES telling FATES how many PFTs to expect on the surface dataset?

**Comment by rosiealice** [comment]
@rgknox - as we were discussing yesterday, I implemented the new hlm_pft_map parameter onto my 'new_fates_nocomp branch'. where I'm working on the no competition modes.
https://github.com/NGEET/fates/compare/master...rosiealice:new_fates_nocomp

The commit with the code for the parameter addition is here: 
https://github.com/rosiealice/fates/commit/a69ac89051a57b55f453d079a01550f3ea661d76

The changes to the parameter file are to be found in this commit:
https://github.com/rosiealice/fates/commit/f0170fd2834a1db26d83df0b9d0e3b0c9e3adcb5

To recap, this s a 2D array that allows for conditions when we want to use the HLM surface dataset to specify the geographical coverage of PFTs, but don't want to use exactly the same PFTs.  So, we can aggregate PFTs where the number of FATES PFTs is less than the HLM, or split a single HLM area (sat, BETs) into lots of seperate FATES PFTs (say, ones with different hydraulic strategies)...

**Comment by rosiealice** [comment]
This is now part of the SP PR and I think it works fine...
